internal/prefixcollector/pod_monitor_server.go

```diff
@@ -0,0 +1,295 @@
+		if err != nil {
+			return nil, err
+		}
+		return ipNet, nil
+	}
+
+	return WatchSubnet(nodeWatcher, keyFunc, subnetFunc)
+}
+
+func watchServiceIPAddr(cs kubernetes.Interface) (*SubnetWatcher, error) {
+	serviceWatcher, err := newServiceWatcher(cs)
+	if err != nil {
+		logrus.Error(err)
+		return nil, err
+	}
+
+	serviceCast := func(obj runtime.Object) (*v1.Service, error) {
+		service, ok := obj.(*v1.Service)
+		if !ok {
+			return nil, errors.Errorf("Casting object to *v1.Service failed. Object: %v", obj)
+		}
+		return service, nil
+	}
+
+	keyFunc := func(event watch.Event) (string, error) {
+		service, err := serviceCast(event.Object)
+		if err != nil {
+			return "", err
+		}
+		return service.Name, nil
+	}
+
+	subnetFunc := func(event watch.Event) (*net.IPNet, error) {
+		service, err := serviceCast(event.Object)
+		if err != nil {
+			return nil, err
+		}
+		ipAddr := service.Spec.ClusterIP
+		return ipToNet(net.ParseIP(ipAddr)), nil
+	}
+
+	return WatchSubnet(serviceWatcher, keyFunc, subnetFunc)
+}
+
+func ipToNet(ipAddr net.IP) *net.IPNet {
+	mask := net.CIDRMask(len(ipAddr)*8, len(ipAddr)*8)
+	return &net.IPNet{IP: ipAddr, Mask: mask}
+}
+
+type serviceWatcher struct {
+	resultCh chan watch.Event
+	stopCh   chan struct{}
+}
+
+func (s *serviceWatcher) Stop() {
+	close(s.stopCh)
+}
+
+func (s *serviceWatcher) ResultChan() <-chan watch.Event {
+	return s.resultCh
+}
+
+func newServiceWatcher(cs kubernetes.Interface) (watch.Interface, error) {
+	ns, err := getNamespaces(cs)
+	if err != nil {
+		return nil, err
+	}
+	resultCh := make(chan watch.Event, 10)
+	stopCh := make(chan struct{})
+
+	for _, n := range ns {
+		w, err := cs.CoreV1().Services(n).Watch(context.TODO(), metav1.ListOptions{})
+		if err != nil {
+			logrus.Errorf("Unable to watch services in %v namespace: %v", n, err)
+			close(stopCh)
+			return nil, err
+		}
+
+		go func() {
+			for {
+				select {
+				case <-stopCh:
+					return
+				case e := <-w.ResultChan():
+					resultCh <- e
+				}
+			}
+		}()
+	}
+
+	return &serviceWatcher{
+		resultCh: resultCh,
+		stopCh:   stopCh,
+	}, nil
+}
+
+func getNamespaces(cs kubernetes.Interface) ([]string, error) {
+	ns, err := cs.CoreV1().Namespaces().List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	rv := []string{}
+	for i := range ns.Items {
+		rv = append(rv, ns.Items[i].Name)
+	}
+	return rv, nil
+}
+
+// WatchSubnet waits for subnets from resourceWatcher, gets subnetwork from watch.Event using subnetFunc.
+// All subnets received from resourceWatcher will be forwarded to subnetCh of returned SubnetWatcher.
+func WatchSubnet(resourceWatcher watch.Interface, keyFunc keyFunc, subnetFunc subnetFunc) (*SubnetWatcher, error) {
+	subnetCh := make(chan *net.IPNet, 10)
+	stopCh := make(chan struct{})
+
+	cache := map[string]string{}
+	var lastIPNet *net.IPNet
+
+	go func() {
+		for {
+			select {
+			case <-stopCh:
+				resourceWatcher.Stop()
+				return
```

Comment:
It looks like we can remove this stop chan logic for both watchers and just use <-ctx.Done in select

Reply:
Removed.